src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+      tempo: options.tempo,
+      textToSpeechLanguage: options.textToSpeechLanguage,
+      videoState: options.videoState,
+      videoTransparency: options.videoTransparency
+    };
+  }
+
+  function serializeProject(project: Project): sb3.ProjectJSON {
+    // Serialize a project. This is the master function used when project.toSb3
+    // is called. The main purpose of serializeProject is to serialize each
+    // target (sprite or stage) and collect them together in the final output
+    // format. It also provides utility functions shared across every target's
+    // serialization, e.g. broadcast utilities.
+
+    // Maintain broadcast dictionaries in both name <-> id directions. sb-edit
+    // operates on broadcast names, but Scratch 3.0 largely requires access to
+    // their IDs as well, so we generate IDs for names as they are required.
+    // Name -> ID is for converting sb-edit names and is the backbone behind
+    // getBroadcastId; ID -> name is the format which is directly serialized
+    // on the stage object. (Only the stage need carries a filled-out broadcast
+    // dictionary in Scratch 3.0.)
+    let broadcastNameToId: { [name: string]: string } = {};
+    let broadcastIdToName: { [id: string]: string } = {};
+
+    const getBroadcastId = (name: string): string => {
+      if (!(name in broadcastNameToId)) {
+        const id = generateId();
+        broadcastNameToId[name] = id;
+        broadcastIdToName[id] = name;
+      }
+      return broadcastNameToId[name];
     };
-    return sb3Sprite;
-  });
+
+    // Preemptively parse through all the broadcast inputs in the project and
+    // generate a broadcast ID for each of them.
+    for (const target of [project.stage, ...project.sprites]) {
+      for (const block of target.blocks) {
+        if (block.opcode === OpCode.event_whenbroadcastreceived) {
+          if (block.inputs.BROADCAST_OPTION.type === "broadcast") {
+            getBroadcastId(block.inputs.BROADCAST_OPTION.value);
+          }
+        } else if (block.opcode === OpCode.event_broadcast || block.opcode === OpCode.event_broadcastandwait) {
+          if (block.inputs.BROADCAST_INPUT.type === "broadcast") {
+            getBroadcastId(block.inputs.BROADCAST_INPUT.value);
+          }
+        }
+      }
+    }
+
+    // Sort broadcasts by name.
+    broadcastNameToId = Object.assign(
+      {},
+      ...Object.entries(broadcastNameToId)
+        .sort(([name1], [name2]) => (name1 < name2 ? -1 : 1))
+        .map(([name, id]) => ({ [name]: id }))
+    );
+    broadcastIdToName = Object.assign(
+      {},
+      ...Object.entries(broadcastIdToName)
+        .sort(([, name1], [, name2]) => (name1 < name2 ? -1 : 1))
+        .map(([id, name]) => ({ [id]: name }))
+    );
+
+    // Set the broadcast name used in obscured broadcast inputs to the first
+    // sorted-alphabetically broadcast's name.
+    getBroadcastId.initialBroadcastName = Object.keys(broadcastNameToId)[0] || "message1";
```

Comment:
Object key order is, in some cases, [not guaranteed by the ECMAScript spec](https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties), so it may be better to use some other data structure if you want to guarantee a specific sort order.

In particular, it looks like here, you're only sorting these objects so that you can set `initialBroadcastName` to the one that comes alphabetically "first", and so you could easily replace the sorting altogether with a "find alphabetically lowest" method.

Reply:
Also, modifying a function's prototype to add a custom property seems kinda janky